_posts/2016-06-30-state-in-scalable-architectures.md

```diff
@@ -0,0 +1,173 @@
+
+* Session scoped: HTTP is a stateless protocol. Remember that *time* is the key word in our definition of state. That means that HTTP doesn't have any built-in mechanism to keep track of state. The concept of session involves a logical group of HTTP requests with some internal state. Classic solution was keeping that state in memory on the service. This solution goes against horizontal scalability. The main point of having stateless services is being able to route requests to different services depending on health and current load of those services. That's not possible when every single service is coupled with the clients through that session state.
+
+* Stream scoped: this is the domain of near real time analytics. State is scoped for a window of time to calculate some analytics such as average, count or max.
+
+* Durable scoped: some data will stand even longer than the code that created it. In order to be sure about its durability we need to store it in some disk based engine like [Cassandra](http://cassandra.apache.org/), [Hadoop](http://hadoop.apache.org/) or [S3](https://aws.amazon.com/documentation/s3/).
+
+Let's focus on state that should be stored for an undefined period of time.
+
+### Database as single source of state
+
+Our next source of inspiration will be [Making sense of Stream Processing](http://www.confluent.io/making-sense-of-stream-processing-ebook-download) ebook. Martin Kleppmann does a portrait of the increasing complexity when using a classic database as the Single Source of Truth, SSOT.
+
+We said before that avoiding state as much as possible is a good idea. Avoiding is a misleading verb, so don't get the impression that a stateless service doesn't have to cope with state, it just delegates it to a data store, expert on that non-functional domain (durability, isolation, consistency, replication...). Same with Serverless architectures, of course there is a server, but you don't have to operate it on that paradigm, you delegate it to experts.
+
+Delegating our durable state to a database is good idea then. However what happens when our system starts to slow down as the load increases? We could then decide to include a cache. Same happens if we need to have full text search capabilities and our DB is not able to cope with that efficiently. Suddenly, we have a SSOT, with at least two projections of that state, optimised for different query patterns. We have two options to keep those views of the state synchronised.
+
+### Database as backbone for data integration
+
+As we said, databases are experts on their own domain. Data integration, with different sources and sinks, lands outside of that domain. We'd like to use the same code that we use for operational transformations (with the same testing and deployment processes) to integrate data. Let's see how we can solve this issue using application code.
+
+### Dual writes for data integration
```

Comment:
Dual writes is a [well known term](https://www.google.es/search?sourceid=chrome-psyapi2&ion=1&espv=2&ie=UTF-8&q=dual%20writes&oq=dual%20writes&aqs=chrome..69i57j0.2071j0j7) in that domain, so changing to concurrent writes would be a little bit confusing.